Define the constructor for AudioNodes (#1117)

padenot opened the pull request:
The first commit is the most important of this PR, it defines what happens when constructing AudioNodes (either via the factory method or the ctor). The second fixes the descriptions of the various factory method to link back to the algorithm used (this is mostly a mechanical change). The third patch fixes a mistake in the first patch, and the fourth one defines the constructors in terms of all the above, linking back to the definitions appropriately: it's mostly mechanical, but adjustments have been performed for the AudioWorklet.

rtoy commented on index.html:
```diff
@@ -2360,68 +2363,114 @@ <h2 id="OfflineAudioCompletionEvent">
+          in, execute these steps, with <var>k</var> the key of the member, and
+          <var>v</var> its value:
+          <ol>
+            <li>If<var>k</var> is <code>disableNormalization</code> or
```
s/If/If /

rtoy commented on index.html:
```diff
@@ -4265,9 +4314,32 @@ <h2 id="GainNode">
+              optional GainOptions options
+            </dt>
+            <dd>
+              An optional initial gain value for this <a>GainNode</a>.
```
|options| isn't the actual initial gain value. Not sure how best to describe this.  Maybe say nothing at all since there's a link to the description of the actual GainOptions dict?  Same goes for all of the following FooOptions dictionaries.

rtoy commented on index.html:
```diff
@@ -2360,69 +2367,124 @@ <h2 id="OfflineAudioCompletionEvent">
+        </ol>
+        <p>
+          <dfn id="audionode-constructor-common">Initializing the common
+          part</dfn> of an object <var>o</var> of interface <var>i</var> that
```
This talks about initializing the common parts like the number of inputs/outputs and channel parameters, and it does.  But then it also talks about the non-common parts like the node-specific options.

I think this needs to be split up into two pieces or just renamed to so that's not about initializing the common part.

padenot replied:
I've renamed it, you're right.

rtoy commented on index.html:
```diff
@@ -8950,6 +9369,31 @@ <h3 id="example-2">
+                be <a href="#associated">associated</a> with.
+              </dd>
+              <dt>
+                optional IIRFilterOptions options
```
This isn't optional.

padenot replied:
Fixed

rtoy commented on index.html:
```diff
@@ -9744,7 +10240,35 @@ <h2 id="MediaStreamAudioSourceNode">
+                "#associated">associated</a> with.
+              </dd>
+              <dt>
+                optional MediaStreamAudioSourceOptions options
```
This isn't optional.

padenot replied:
Fixed

padenot commented on index.html:
```diff
@@ -2360,69 +2367,124 @@ <h2 id="OfflineAudioCompletionEvent">
+          <li>Let <var>o</var> be a new object of type <var>i</var>.
+          </li>
+          <li>Let <var>option</var> be a dictionnary of the type associated to
+          the interface associated to this factory method.
```
Done. I also linked more things.

padenot commented on index.html:
```diff
@@ -4269,6 +4331,31 @@ <h2 id="GainNode">
         "[Constructor(&lt;a&gt;BaseAudioContext&lt;/a&gt; context, optional &lt;a&gt;GainOptions&lt;/a&gt; options)]interface GainNode : &lt;a&gt;AudioNode&lt;/a&gt;"
         class="idl">
```
apparently when you define it in the `<dl>`, then you don't have to do it in the title. I'll fix.

padenot commented on index.html:
```diff
@@ -4269,6 +4331,31 @@ <h2 id="GainNode">
+            <p>
+              Let <var>gain</var> be a new <a>GainNode</a> object. <a href=
+              "#audionode-constructor-common">Initialize the common part</a> of
+              <var>gain</var>, and return <var>gain</var>.
```
This is not called "initializing an AudioNode", hopefully less confusing.

padenot commented on index.html:
```diff
@@ -2360,69 +2367,124 @@ <h2 id="OfflineAudioCompletionEvent">
+          in <var>dict</var>, and then set its <a href=
+          "#widl-ConvolverNode-buffer">buffer</a> attribute to the value of the
+          <a href="#widl-ConvolverOptions-buffer">buffer</a> in <var>dict</var>
+          member, in this order, and jump to the last step of this algorithm.
```
Yes, this algorithm describes a way of getting the desired result. There are other ways to get the same result.

padenot commented on index.html:
```diff
@@ -2360,69 +2367,124 @@ <h2 id="OfflineAudioCompletionEvent">
-              method</a> function for this <a>AudioNode</a> type. Initialize
-              all elements of this array to <code>null</code>.
+              <li>If <var>k</var> is <code>disableNormalization</code> or
+              <code>buffer</code>, jump to the beginning of this loop.
```
Fixed.

padenot commented:
@rtoy, sorry for the delays, I've addressed the issues you found, preview [here](https://rawgit.com/padenot/web-audio-api/a338555737720fc5002ec137a3b7ab8669a33db7/index.html).

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+        </p>
+        <section>
+          <h3>
+            AudioNode creation
```
Nit: "creation" should be capitalized.

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+            "associated">associated <a>BaseAudioContext</a></dfn> of the
+            <a>AudioNode</a> to be created. Similarly, when using the factory
+            method, the <a>associated <code>BaseAudioContext</code></a> of the
+            <a>AudioNode</a> is the first argument of the factory method.
```
This seems wrong since many factory methods don't have a first argument at all.  Or perhaps I don't understand what you're trying to say here.

padenot replied:
Yeah this is wrong. I meant "The associated BaseAudioContext when calling a factory method is the BaseAudioContext this factory method is called on", which makes sense.

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+            <a>AudioNode</a> is the first argument of the factory method.
+          </p>
+          <p>
+            To create a new <a>AudioNode</a> of a particular type <var>i</var>
```
Can we use "n" instead of "i"? I find it hard to distinguish between an "i" and a "l" and a "1".

padenot replied:
Done

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+            </li>
+            <li>
+              <a href="#audionode-constructor-init">Initialize the common
+              part</a> of <var>o</var>, with <var>c</var> and <var>option</var>
```
I assume you don't want to say "Initializing the common part" here anymore because that section isn't about just the common part anymore.

padenot replied:
Yeah I half assed my regex apparently. Fixed.

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+          <ol>
+            <li>Let <var>o</var> be a new object of type <var>i</var>.
+            </li>
+            <li>Let <var>option</var> be a dictionnary of the type <a href=
```
Typo: "dictionnary" -> "dictionary"

rtoy commented on index.html:
```diff
@@ -2363,68 +2370,131 @@ <h2 id="OfflineAudioCompletionEvent">
+            value of this parameter.
+            </li>
+            <li>
+              <a href="#audionode-constructor-init">Initialize the common
```
Ditto. It's not the common part anymore.

rtoy commented on index.html:
```diff
@@ -4430,9 +4548,7 @@ <h2 id="AudioBuffer">
-        <dl title=
-        "[Constructor(&lt;a&gt;AudioBufferOptions&lt;a&gt; options)]interface AudioBuffer"
-        class="idl">
+        <dl title="interface AudioBuffer" class="idl">
```
I think you forget to define the actual constructor for this.  And the constructor for this doesn't follow the AudioNode constructors because context is not the first arg.  Not sure where the best place to describe the constructor should be.

padenot replied:
Those are not `AudioNode`s, the goal is to fix it in a different PR (this PR and issue is about AudioNodes, there are other issues about other objects).

This is different because the notion of "associated AudioContext" does not apply, among other things. I've reinstated this constructor.

Anyways, I plan on following up immediately with defining the ctors for other objects, this PR is big enough already.

rtoy replied:
sgtm.

rtoy commented on index.html:
```diff
@@ -5096,8 +5265,36 @@ <h2 id="MediaElementAudioSourceNode">
+                "#associated">associated</a> with.
+              </dd>
+              <dt>
+                optional MediaElementAudioSourceOptions options
```
This is not optional.

rtoy commented on index.html:
```diff
@@ -5096,8 +5265,36 @@ <h2 id="MediaElementAudioSourceNode">
+              </dd>
+            </dl>
+          </dd>
+        </dl>
```
Something is messed up somewhere.  The Media* nodes all used to be in successive sections.  Now the worklet node comes after MediaElementAudioSource, and the other nodes are all placed at the end.

padenot replied:
This PR does not change this, but I can do a patch to reorder the section if you'd like.

rtoy replied:
If you haven't already, don't bother. Recall that we're going to reorder all the sections anyway to roughly alphabetical.

rtoy commented on index.html:
```diff
@@ -8950,10 +9366,34 @@ <h3 id="example-2">
+                IIRFilterOptions options
+              </dt>
+              <dd>
+                Optional initial parameter value for this <a>IIRFilterNode</a>.
```
Not optional.

padenot replied:
Fixed.

rtoy commented:
For those following along:  https://rawgit.com/WebAudio/web-audio-api/591f0a7/index.html
